docs/dev-guide/contrib-guides-cpp.md

```diff
@@ -0,0 +1,158 @@
+it.
+
+## Guidelines
+
+We adhere to [Google's C++ style guide][google-cpp-style-guide] (as of
+[8f97e24][google-styleguide-8f97e24]) with the following exceptions (organized according to the
+sections in Google's style guide).
+
+:::{note}
+This section is a work in progress and does not yet include all exceptions after the
+[Classes][google-cpp-style-guide-classes] section of Google's style guide.
+:::
+
+### Header files
+
+#### Self-contained Headers
+
+* Header files should end in `.hpp`.
+* Don't use non-header files meant for inclusion, since they can confuse static analysis tools.
+
+#### The #define Guard
+
+The symbol name should have the form
+
+```txt
+<NAMESPACE>_<FILENAME-STEM>_<FILENAME-EXTENSION>
+```
+
+where
+
+* `<NAMESPACE>` is the namespace of the file.
+  * For files in a nested namespace, each namespace layer should be separated by an underscore.
+* `<FILENAME_STEM>` is the file's name without the extension.
+  * For stems with multiple words, the words should _not_ be separated with underscores.
+* `<FILENAME_EXTENSION>` is the file's extension.
+
+For example:
+
+`clp/streaming_archive/reader/SegmentManager.hpp`
+
+```c++
+#ifndef CLP_STREAMING_ARCHIVE_READER_SEGMENTMANAGER_HPP
+#define CLP_STREAMING_ARCHIVE_READER_SEGMENTMANAGER_HPP
+
+namespace clp::streaming_archive::reader {
+    // ...
+}
+
+#endif  // CLP_STREAMING_ARCHIVE_READER_SEGMENTMANAGER_HPP
+```
+
+#### Names and order of includes
+
+* For codebases where the code is not organized into well-defined libraries, it is fine to use UNIX
+  directory aliases to include headers.
+* For C headers that have C++ counterparts (e.g., `stddef.h` vs `cstddef`), use the C++ counterpart.
+
+### Scoping
+
+#### Namespaces
+
+Single-line nested namespace declarations (e.g. `bar` in `foo`) should use the following format
+(unless doing so would affect clarity):
+
+```cpp
+namespace foo::bar {
+}
+```
+
+#### Internal linkage
+
+Only use unnamed namespaces (instead of the `static` qualifier) to give functions and variables
+internal linkage. However, as Google's style guide indicates, you can't use unnamed namespaces in
+header files. For symbols that should only be used within a header file, you can create a named
+namespace in the header file, where its name is of the form:
+
+```txt
+<FilenameStem>_internal
+```
+
+where
+
+* `<FilenameStem_stem>` is the file's name without the extension.
+* `_internal` is a fixed suffix indicating the namespace's purpose.
+
+For example:
+
+`clp/streaming_archive/reader/SegmentManager.hpp`
+
+```c++
+namespace clp::streaming_archive::reader {
+namespace SegmentManager_internal {
+// Internal symbols
+}
+}
+```
+
+### Classes
+
+#### Doing work in constructors
+
+We allow (but discourage) the use of exceptions, even in constructors. If creating an object can
+fail, you're encouraged to use a factory function that performs the work that can fail, and if
+successful, constructs and returns the object.
```

Comment:
Should we encourage people to return a result here? If someone explicitly follows this and returns an object you can't do much better error handling than a constructor, no?

Reply:
True, made it clearer.